server/sonar-web/src/main/js/apps/users/components/TokensForm.tsx

```diff
@@ -0,0 +1,196 @@
+    getTokens(user.login).then(
+      tokens => {
+        if (this.mounted) {
+          this.setState({ loading: false, tokens });
+        }
+      },
+      () => {
+        if (this.mounted) {
+          this.setState({ loading: false });
+        }
+      }
+    );
+  };
+
+  handleCloseClick = (evt: React.SyntheticEvent<HTMLAnchorElement>) => {
+    evt.preventDefault();
+    this.handleClose();
+  };
+
+  handleClose = () => {
+    if (this.state.hasChanged) {
+      this.props.onUpdateUsers();
+    }
+    this.props.onClose();
+  };
+
+  handleGenerateToken = (evt: React.SyntheticEvent<HTMLFormElement>) => {
+    evt.preventDefault();
+    if (this.state.newTokenName.length > 0) {
+      this.setState({ generating: true });
+      generateToken({ name: this.state.newTokenName, login: this.props.user.login }).then(
+        newToken => {
+          if (this.mounted) {
+            this.fetchTokens();
+            this.setState({ generating: false, hasChanged: true, newToken, newTokenName: '' });
+          }
+        },
+        () => {
+          if (this.mounted) {
+            this.setState({ generating: false });
+          }
+        }
+      );
+    }
+  };
+
+  handleRevokeToken = () => {
+    this.setState({ hasChanged: true });
+    this.fetchTokens();
+  };
+
+  handleNewTokenChange = (evt: React.SyntheticEvent<HTMLInputElement>) =>
+    this.setState({ newTokenName: evt.currentTarget.value });
+
+  render() {
+    const { generating, loading, newToken, newTokenName, tokens } = this.state;
+    const header = translate('users.tokens');
+
+    return (
+      <Modal contentLabel={header} onRequestClose={this.handleClose}>
+        <header className="modal-head">
+          <h2>{header}</h2>
+        </header>
+        <div className="modal-body modal-container">
+          <h3 className="spacer-bottom">{translate('users.generate_tokens')}</h3>
+          <form id="generate-token-form" onSubmit={this.handleGenerateToken} autoComplete="off">
+            <input
+              className="spacer-right"
+              type="text"
+              maxLength={100}
+              onChange={this.handleNewTokenChange}
+              placeholder={translate('users.enter_token_name')}
+              required={true}
+              value={newTokenName}
+            />
+            <button
+              className="js-generate-token"
+              disabled={generating || newTokenName.length <= 0}
+              type="submit">
+              {translate('users.generate')}
+            </button>
+          </form>
+
+          {newToken && <TokensFormNewToken token={newToken} />}
+
+          <table className="data zebra big-spacer-top ">
+            <thead>
+              <tr>
+                <th>{translate('name')}</th>
+                <th className="text-right">{translate('created')}</th>
+                <th />
+              </tr>
+            </thead>
+            <tbody>
+              <DeferredSpinner
```

Comment:
this `DeferredSpinner` is hard to read, could you extract its parts to dedicated variables/functions?